pkg/ioutils/bytespipe.go

```diff
@@ -0,0 +1,93 @@
+package ioutils
+
+const maxCap = 1e6
+
+// BytesPipe is io.ReadWriter which works similary to pipe(queue).
+// All written data could be read only once. Also BytesPipe trying to adjust
+// internal []byte slice to current needs, so there won't be overgrown buffer
+// after highload peak.
+// BytesPipe isn't goroutine-safe, caller must synchronize it if needed.
+type BytesPipe struct {
+	buf        [][]byte // slice of byte-slices of buffered data
+	lastRead   int      // index in the first slice to a read point
+	bufLen     int      // length of data buffered over the slices
+	bufCap     int      // whole allocated capacity over slices
+	writeIndex int      // index to the buffer where the writes go
+}
+
+// NewBytesPipe creates new BytesPipe, initialized by specified slice.
+// If buf is nil, then it will be initialized with slice which cap is 64.
+// buf will be adjusted in a way that len(buf) == 0, cap(buf) == cap(buf).
+func NewBytesPipe(buf []byte) *BytesPipe {
+	if cap(buf) == 0 {
+		buf = make([]byte, 0, 64)
+	}
+	return &BytesPipe{
+		buf:    [][]byte{buf[:0]},
+		bufCap: cap(buf),
+	}
+}
+
+func (bp *BytesPipe) grow(n int) {
+	for bp.bufLen+n > bp.bufCap {
+		// not enough space
+		nextCap := 2 * cap(bp.buf[len(bp.buf)-1])
+		if maxCap < nextCap {
+			nextCap = maxCap
+		}
+		bp.buf = append(bp.buf, make([]byte, 0, nextCap))
+		bp.bufCap += nextCap
+	}
+}
+
+// Write writes p to BytesPipe.
+// It can increase cap of internal []byte slice in a process of writing.
+func (bp *BytesPipe) Write(p []byte) (n int, err error) {
+	bp.grow(len(p))
```

Comment:
Not really sure why there are two separate loops involved here (one to allocate space, and another to copy). I think the code would be more concise and readable if this happened in the same loop.